miner/scroll_worker.go

```diff
@@ -0,0 +1,943 @@
+			// and processed first in every block anyways
+			w.prioritizedTx = &prioritizedTransaction{
+				blockNumber: w.currentPipeline.Header.Number.Uint64() + 1,
+				tx:          res.OverflowingTx,
+			}
+		}
+
+		switch {
+		case res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrBlockRowConsumptionOverflow):
+			l1TxRowConsumptionOverflowCounter.Inc(1)
+		case !res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrBlockRowConsumptionOverflow):
+			l2TxRowConsumptionOverflowCounter.Inc(1)
+		case res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l1TxCccUnknownErrCounter.Inc(1)
+		case !res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l2TxCccUnknownErrCounter.Inc(1)
+		}
+	}
+
+	if !w.isRunning() {
+		if res != nil && res.FinalBlock != nil {
+			w.updateSnapshot(res.FinalBlock)
+		}
+		w.currentPipeline = nil
+		return nil
+	}
+
+	if res == nil || res.FinalBlock == nil {
+		w.startNewPipeline(time.Now().Unix())
+		return nil
+	}
+	return w.commit(res)
+}
+
+// commit runs any post-transaction state modifications, assembles the final block
+// and commits new work if consensus engine is running.
+func (w *worker) commit(res *pipeline.Result) error {
+	defer func(t0 time.Time) {
+		l2CommitTimer.Update(time.Since(t0))
+	}(time.Now())
+
+	if res.CCCErr != nil {
+		commitReasonCCCCounter.Inc(1)
+	} else {
+		commitReasonDeadlineCounter.Inc(1)
+	}
+
+	block, err := w.engine.FinalizeAndAssemble(w.chain, res.FinalBlock.Header, res.FinalBlock.State,
+		res.FinalBlock.Txs, nil, res.FinalBlock.Receipts)
+	if err != nil {
+		return err
+	}
+
+	select {
+	case w.taskCh <- &task{receipts: res.FinalBlock.Receipts, state: res.FinalBlock.State, block: block, createdAt: time.Now(),
+		accRows: res.Rows, nextL1MsgIndex: res.FinalBlock.NextL1MsgIndex}:
+		w.unconfirmed.Shift(block.NumberU64() - 1)
+		log.Info("Commit new mining work", "number", block.Number(), "sealhash", w.engine.SealHash(block.Header()),
+			"txs", res.FinalBlock.Txs.Len(),
+			"gas", block.GasUsed(), "fees", totalFees(block, res.FinalBlock.Receipts),
+			"elapsed", common.PrettyDuration(time.Since(w.currentPipelineStart)))
+	case <-w.exitCh:
+		log.Info("Worker has exited")
+	}
+
+	w.currentPipeline = nil
+	return nil
+}
+
+// copyReceipts makes a deep copy of the given receipts.
+func copyReceipts(receipts []*types.Receipt) []*types.Receipt {
+	result := make([]*types.Receipt, len(receipts))
+	for i, l := range receipts {
+		cpy := *l
+		result[i] = &cpy
+	}
+	return result
+}
+
+// postSideBlock fires a side chain event, only use it for testing.
+func (w *worker) postSideBlock(event core.ChainSideEvent) {
+	select {
+	case w.chainSideCh <- event:
+	case <-w.exitCh:
+	}
+}
+
+func (w *worker) onTxFailingInPipeline(txIndex int, tx *types.Transaction, err error) bool {
```

Comment:
what does this return value indicate? it seems always returning false

Reply:
It allows caller to terminate the pipeline early when a txn fails apply stage. https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-3236476391a7611535994ed7b11821c5e7b6cf6056e0a49c498dbfb1e249f43dR129

I will document it better.